### PR TITLE
Some minor syntax API changes for patterns

### LINKED ISSUE #43452: naming inconsistency between BinaryPattern and BinaryExpression
BinaryExpression is  Left/OperatorToken/Right
BinaryPattern is LeftPattern/PatternOperator/RightPattern

We should unify the latter on hte former.

We should also check UnaryPattern vs UnaryExpression.

### LINKED ISSUE #43453: Consider providing a common base type between ExpressionSyntax and PatternSyntax
IDE has a lot of codepaths that want to treat these uniformly.  i.e. for purposes of things like parenthesization (and the reverse), precedence and the like.  With the introduction of patterns (and the fact that patterns and expression bridge through IsPatternExpression (which bridges from expression to pattern) and ConstantPattern (which bridges from pattern to expression), a lot of our codepaths have gotten very degraded since the only type we can use in common is effectively CSharpSyntaxNode.

This means that we effectively have no guardrails anywhere here, and we have to annotate all these methods with asserts/contracts that we either have a Pattern or Expression.  And it's still easy to potentially fall off when people just blindly pass around something like `.Parent` (which would have previously errored when things were strongly typed, forcing the caller to think about it).

I really don't love ExpressionOrPatternSyntax.  But it would provdie a way for us to be much more strongly typed here and to prevent accidental problems that are easy when everything just takes SyntaxNode.

### PR DESCRIPTION
**Pull request by gafter**
- Change name to OperatorToken in UnaryPattern and BinaryPattern.
Fixes #43452
- Add ExpressionOrPatternSyntax as a common base type.
Fixes #43453

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/AddRequiredParentheses/CSharpAddRequiredPatternParenthesesDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -42,7 +42,7 @@ protected override bool IsBinaryLike(PatternSyntax node)
         protected override (PatternSyntax, SyntaxToken, PatternSyntax) GetPartsOfBinaryLike(BinaryPatternSyntax binaryPattern)
         {
             Debug.Assert(IsBinaryLike(binaryPattern));
-            return (binaryPattern.LeftPattern, binaryPattern.PatternOperator, binaryPattern.RightPattern);
+            return (binaryPattern.LeftPattern, binaryPattern.OperatorToken, binaryPattern.RightPattern);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/AddRequiredParentheses/CSharpAddRequiredPatternParenthesesDiagnosticAnalyzer.cs:45]: didn't we want LeftPattern and RightPattern to be renamed to Left/Right to also match BinaryExpression?

Comment (gafter) [reply src/Analyzers/CSharp/Analyzers/AddRequiredParentheses/CSharpAddRequiredPatternParenthesesDiagnosticAnalyzer.cs:45]: > didn't we want LeftPattern and RightPattern to be renamed to Left/Right to also match BinaryExpression?

This is the first time I'm aware of that suggestion. @dotnet/roslyn-compiler Any opinions?

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/AddRequiredParentheses/CSharpAddRequiredPatternParenthesesDiagnosticAnalyzer.cs:45]: Sorry, i thought it was clear in https://github.com/dotnet/roslyn/issues/43452  :)

Yes, i'm asking for BinaryPattern/Expr to be uniform here in naming for all the children.  Thanks!

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 2]
```diff
@@ -1881,7 +1881,8 @@
     </Field>
     <Field Name="Condition" Type="ExpressionSyntax"/>
   </Node>
-  <AbstractNode Name="PatternSyntax" Base="CSharpSyntaxNode" />
+  <AbstractNode Name="ExpressionOrPatternSyntax" Base="CSharpSyntaxNode" />
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml:1884]: ><AbstractNode Name="ExpressionOrPatternSyntax" Base="CSharpSyntaxNode" /> [](start = 2, length = 73)

Consider placing this node above all derived nodes.

File: src/Analyzers/CSharp/Analyzers/AddRequiredParentheses/CSharpAddRequiredPatternParenthesesDiagnosticAnalyzer.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Analyzers/CSharp/Analyzers/AddRequiredParentheses/CSharpAddRequiredPatternParenthesesDiagnosticAnalyzer.cs:45]: I also assumed this is part of the proposal in #43452 and commented there with that in mind.

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 3]
```diff
@@ -1985,15 +1986,15 @@
...
     </Field>
     <Field Name="RightPattern" Type="PatternSyntax" />
   </Node>
   <Node Name="UnaryPatternSyntax" Base="PatternSyntax">
     <Kind Name="NotPattern"/>
-    <Field Name="PatternOperator" Type="SyntaxToken">
+    <Field Name="OperatorToken" Type="SyntaxToken">
       <Kind Name="NotKeyword"/>
     </Field>
     <Field Name="Pattern" Type="PatternSyntax" />
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml:2000]: >Pattern [](start = 17, length = 7)

"Operand" for consistency with unary expressions?

**Comment by AlekseyTs** [comment]
It looks like this PR only partially addresses #43452. Consider adjusting the PR description and keeping the issue active after merge.

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml (hunk 3 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml:2000]: I don't think so.  The analogy with expressions breaks down here.  The unary pattern is not an operation, so it doesn't really have an operand.

---
In reply to: [412377963](https://github.com/dotnet/roslyn/pull/43526#discussion_r412377963) [](ancestors = 412377963)

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml:2000]: i'm ok with these being different as well.  it was the binary ones that felt super weird.

**Comment by gafter** [comment]
Reviewers: I think this completes the set of requested API changes. Can you please check the latest changes in iteration 3?

**Pull request merged**